aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Merger.java

```diff
@@ -66,6 +67,18 @@
      * @return A completable operation to merge the model
      */
     <T extends Model> Completable merge(ModelWithMetadata<T> modelWithMetadata) {
+        return merge(modelWithMetadata, value -> { });
+    }
+
+    /**
+     * Merge an item back into the local store, using a default strategy.
+     * @param modelWithMetadata A model, combined with metadata about it
+     * @param storageItemChangeConsumer A callback invoked when the merge method saves or deletes the model.
+     * @param <T> Type of model
+     * @return A completable operation to merge the model
+     */
+    <T extends Model> Completable merge(ModelWithMetadata<T> modelWithMetadata,
+                                        Consumer<StorageItemChange<T>> storageItemChangeConsumer) {
```

Comment:
Is there anyway we can avoid having these callbacks?

The `Completable` is already an asynchronous signal. Now, this method has two different async signals that have different lifecycles. It's a bit to handle, conceptually.

One possible option, if you need to observe save/deletes on the datastore, is the `observe()` method on the `LocalStorageAdapter`? Hopefully that can be setup internally, without effecting the Merger's contract?

Reply:
That was my previous implementation. However, it proved to be cumbersome from the standpoint of determining the lifecycle of the sync attempt for a given model. 

I was creating an observer before kicking off the sync. That observer kept an in-memory map of counters for each model and applied the following logic to each event received.:

```
// pseudo-code
onChangeEvent (changeEvent) {
 if (changeEvent.initiator == SYNC_ENGINE && isUserModel(changeEvent.model)) {
   incrementCounter(changeEvent.model, changeEvent.changeType)
 } else if (changeEvent.initiator == SYNC_ENGINE && changeEvent.model == LastSyncMetadata) {
  // That means the sync attempt finished for one of the models
  let syncMetadata = changeEvent.eventPayload
  //end of sync attempt for model specified in the 
  emitMetricsFor(syncMetadata.modelName);
 } else {
   NoOp
 }
}
```
While this approach worked and it was nicely decoupled, it felt kind of brittle. It assumes that the initial sync  is the only thing emitting events with initiator set to SYNC_ENGINE. While this is true right now, it's one of those things that could change and break this logic. It also assumes that an update to the LastSyncMetadata table made by SYNC_ENGINE was the end of the sync attempt for the model.

Reply:
If I have bandwidth tomorrow, I'll revisit this one to see what I can come up with. Maybe instead of returning a completable, I can return a Single or a Maybe with the StorageItemChange<T>

Reply:
I'll do this as a follow-up.